jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.lambda
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.util.indexing.DataIndexer
+import com.intellij.util.indexing.DefaultFileTypeSpecificInputFilter
+import com.intellij.util.indexing.FileBasedIndex
+import com.intellij.util.indexing.FileBasedIndexExtension
+import com.intellij.util.indexing.FileContent
+import com.intellij.util.indexing.ID
+import com.intellij.util.io.DataExternalizer
+import com.intellij.util.io.EnumeratorStringDescriptor
+import com.intellij.util.io.KeyDescriptor
+import org.jetbrains.yaml.YAMLLanguage
+import org.jetbrains.yaml.psi.YAMLKeyValue
+import software.aws.toolkits.jetbrains.services.cloudformation.IndexedFunction
+import software.aws.toolkits.jetbrains.services.cloudformation.IndexedResource
+import software.aws.toolkits.jetbrains.services.cloudformation.LAMBDA_FUNCTION_TYPE
+import software.aws.toolkits.jetbrains.services.cloudformation.SERVERLESS_FUNCTION_TYPE
+import software.aws.toolkits.jetbrains.services.cloudformation.yaml.YamlCloudFormationTemplate
+import java.io.DataInput
+import java.io.DataOutput
+
+class SamTemplateIndex : FileBasedIndexExtension<String, MutableList<IndexedResource>>() {
+    private val fileFilter by lazy {
+        val supportedFiles = arrayOf(YAMLLanguage.INSTANCE.associatedFileType)
+
+        object : DefaultFileTypeSpecificInputFilter(*supportedFiles) {
+            override fun acceptInput(file: VirtualFile): Boolean = file.isInLocalFileSystem
+        }
+    }
+
+    override fun getValueExternalizer(): DataExternalizer<MutableList<IndexedResource>> = object : DataExternalizer<MutableList<IndexedResource>> {
+        override fun save(dataOutput: DataOutput, value: MutableList<IndexedResource>) {
+            dataOutput.writeInt(value.size)
+            value.forEach { resource ->
+                dataOutput.writeInt(resource.indexedProperties.size)
+                resource.indexedProperties.forEach { key, value ->
+                    dataOutput.writeUTF(key)
+                    dataOutput.writeUTF(value)
+                }
+            }
+        }
+
+        override fun read(dataInput: DataInput): MutableList<IndexedResource> {
+            val size = dataInput.readInt()
```

Comment:
can we call this `resourceCount` - I know it is literally `Collection.size` but it represents the number of resources being read

Reply:
done.